_posts/2021-01-08-consumer-tuning.md

```diff
@@ -0,0 +1,281 @@
+
+## Configuring a reliable data pipeline
+
+To guarantee the reliability of message delivery on the _producer side_, you might configure your producers to use idempotence and transactional ids.
+
+Idempotence
+: Idempotence properties used for idempotence guarantee the order of message delivery from producers for exactly once writes to a single partition.
+
+    ```properties
+    # ...
+    enable.idempotence=true
+    max.in.flight.requests.per.connection=5
+    acks=all
+    retries=2147483647
+    # ...
+    ```
+
+Transactional id
+: Transactional properties guarantee that messages using the same transactional is are produced once, and either all are successfully written to the respective logs or none of them are. The timeout set a time limit on achieving this.
+
+    ```properties
+    # ...
+    transactional.id=_UNIQUE-ID_
+    transaction.timeout.ms=900000
+    # ...
+    ```
+
+With producers set up in such a way, you can make the pipeline more secure from the consumer side by introducing the `isolation.level` property. The `isolation.level` property controls how transactional messages are read by the consumer, and has two valid values:
+
+* `read_committed`
+* `read_uncommitted` (default)
+
+If you switch from the default to `read_committed`, only transactional messages that have been committed are read by the consumer. As always, there's a trade-off. Using this mode will lead to an increase in end-to-end latency, because the consumer will only return a message when the brokers have written the transaction markers that record the result of the transaction (_committed_ or _aborted_).
+
+```properties
+# ...
+enable.auto.commit=false
+isolation.level=read_committed
+# ...
+```
+
+## Recovering from failure within a consumer group
+
+You can define how often checks are made on the health of consumers within a consumer group. If consumers fail within a consumer group, a rebalance is triggered and partition ownership is reassigned to the members of the group. You want to get the timings of your checks just right so that the consumer group can recover quickly, but unnecessary rebalances are not triggered. And you use two properties to do it: `session.timeout.ms` and `heartbeat.interval.ms`.
+
+`session.timeout.ms`
+: Specifies the maximum amount of time in milliseconds a consumer within a consumer group can be out of contact with a broker before being considered inactive and a _rebalancing_ is triggered between the active consumers in the group. Setting the `session.timeout.ms` property lower means failing consumers are detected earlier, and rebalancing can take place quicker. However, you don't want to set the timeout so low that the broker fails to receive an _all-clear_ heartbeat in time and triggers an unnecessary rebalance.
```

Comment:
What is an _all-clear_ heartbeat. I've never seen this term before.

Reply:
Removed!